enhancements/rhcos/split-rhcos-into-layers.md

```diff
@@ -0,0 +1,467 @@
+that can be shared otherwise.
+
+### Implementation Details/Notes/Constraints
+
+As mentioned in the proposal, the work is split into three phases. Here, we go into more
+explicit implementation details for each phase.
+
+##### Phase 1: Using RHEL/CentOS Stream-only bootimages
+
+- In openshift/os, create a new variant of RHCOS which contains solely RHEL packages,
+  i.e. today's RHCOS but without any OCP components.
+  - Note this work has already been done: https://github.com/openshift/os/pull/1445
+- In the RHCOS pipeline, start building this new stream variant. This will push the bootable
+  container to quay.io/openshift-release-dev/ocp-v4.0-art-dev. Bootimages will also be built
+  as usual.
+- Use these bootimages to test changes to the installer boostrapping flow.
+  Notably, bootstrapping currently requires the kubelet and cri-o to be present
+  in order to bring up the temporary control plane. This will be adapted so that
+  the installer first live-applies the node image, as if the node did a system
+  update.
+  - Note this work is already in progress:
+    https://github.com/openshift/installer/pull/8742
+  Other components will also need to be adapted (such as assisted-installer, but
+  likely others too).
+- Work with the teams of the affected components to adapt them.
+- Open an openshift/installer PR which updates the bootimages. This will allow
+  us to run more extensive CI tests on the PR.
+- Raise awareness of this change to the OCP org to invite more testing. 
+- Once all issues have been resolved, merge the PR. If new regressions are
+  encountered, revert the PR and fix them.
+- Once satisfied with the new bootimages, the RHCOS pipeline can stop building
+  the old bootimages containing OpenShift components.
+
+##### Phase 2: Switching `rhel-coreos` in CI
+
+- Set up a new job in openshift/release which builds the layered OpenShift node image on
+  top of the bootable container image from the new stream variant.
+- Open MCO and DTK PRs pointing to this layered image in CI (these are the only two components
+  that explicitly reference `rhel-coreos`) as a way to verify CI is happy.
+    - NOTE: There shouldn't be any major CI failures here. One key thing to understand is that
+      we're changing *how* the node image is built, but not *what* its final contents are from
+      a squashed image perspective. So this is mostly a sanity-check that they are indeed
+      equivalent.
+- Once any issues have been resolved, make the new job push to the `:rhel-coreos` tag of the
+  CI imagestream. This simultaneously requires disabling ART's job which syncs RHCOS to that
+  same tag.
+
+##### Phase 3: Switching `rhel-coreos` in production
+
+- Work with CRT to update the OCP release page to better handle the new layered approach.
+    - It currently links to diffs in the RHCOS release browser. But now the release browser
+      will only contain the RHEL diff, not the OCP diff in the node layer. We should close this
```

Comment:
The release browser would also report the wrong thing if, for example, ART layers on a new kernel. The release controller does not presently have a method of reporting RPM versions without the release browser, so we will lose visibility into the hyperkube version installed in the `node` image from the release controller.

Reply:
"release browser" should go away as a concept: what's in the container image can be found by introspecting the container image.

We must handle the general case of "higher layer overrides/removes content from lower layer" across all containers - the OS is not special.

Reply:
Agreed. But the release controller is stateless and can't introspect all images in a reasonable period of time for a payload details page to render. So I suspect there is significant work to explain here. `oc adm release new` would have the time and access necessary to store the RPM information and it could be new metadata output in `oc adm release info -o json` .

Reply:
AIUI today the diff is dynamic, but I can't think of a problem with us maintaining a long term cache of the extracted rpmdb for each image. I think quay.io may actually do something like this already? The [security scan](https://quay.io/repository/fedora/fedora-bootc/manifest/sha256:7bc5ec286a804d89de2b739725fa82ad6dbb8a17668197d612760ec7ab90ffe3?tab=vulnerabilities) is already has is basically a rendered derivation of that.

Then given that cache, we can compute the diff between two container images easily, which is what the release browser does today.

Reply:
> diff is dynamic

`oc adm release info --changelog` is used. It uses information baked into the release payload to find the commits associated with each of the payload components. `oc` then uses git repo clones to quickly compute the diff based on the commit history.

This changelog output from `oc` is in markdown and the release controller just translates it into HTML. The release controller doesn't presently have any means to interject new elements, like an RPM diff, in the contiguous markdown document without brittle heuristics. 

changelog also doesn't deal with heterogeneous payloads today as commits must agree between all arches of a component. This is not the case for different architectures & RPMs.

So a list of options:
1. Release controller introspects images. Significant change to behavior today. New cache (quay does not have information on registry.ci.openshift.org images -- required for non-OS).
2. `oc` does the introspection. `oc ` includes the RPM data in markdown. quay is not an option because payload component images are not necessarily on quay.io (and it seems unreasonable to require users to have a quay token). New views / choreography with release controller unless RPMs are included inline within the markdown (could be voluminous / difficult to read). 
3. Something else?

#2 makes sense to me. In any case, I feel there are weeks of work here for one or more teams we shouldn't discount if the release browser disappears.

Reply:
I'll distill this conversation in the enhancement, but ISTM we don't need to have it fully fleshed out before getting this merged. But yes, definitely a gap to address before rolling out to prod.

Reply:
Chatted with Justin about this a bit more and did some tests. It's very expensive to extract the rpmdb. First, currently it seems like `oc image extract` chokes on hardlinks so we'd have to fix that. I think one hack would be to always end the node image build with e.g. `RUN rpm -qa > /usr/lib/sysimage/rpm-base.txt` or something, and then `oc image extract` could much more cheaply pull down just the final layer. But yuck...

Anyway, I'm thinking the simplest is probably to just keep using the release browser. Have a service that just caches the rpmdb from built node images and pushes it out to S3 in a way that the release browser already knows how to consume.